File: kotlin/src/AoCPuzzle.kt
import com.github.ajalt.mordant.rendering.TextAlign
import com.github.ajalt.mordant.rendering.TextColors.brightGreen
import com.github.ajalt.mordant.rendering.TextColors.brightRed
import com.github.ajalt.mordant.rendering.TextColors.cyan
import com.github.ajalt.mordant.rendering.TextColors.gray
import com.github.ajalt.mordant.rendering.TextColors.yellow
import com.github.ajalt.mordant.rendering.TextStyle
import com.github.ajalt.mordant.rendering.TextStyles.reset
import com.github.ajalt.mordant.rendering.TextStyles.bold
import com.github.ajalt.mordant.rendering.Whitespace
import com.github.ajalt.mordant.table.ColumnWidth
import com.github.ajalt.mordant.table.grid
import com.github.ajalt.mordant.table.horizontalLayout
import com.github.ajalt.mordant.terminal.Terminal
import com.github.ajalt.mordant.widgets.HorizontalRule
import com.github.ajalt.mordant.widgets.Padding
import util.Input
import util.downloadInput
import kotlin.io.path.Path
import kotlin.io.path.notExists
import kotlin.io.path.readLines
import kotlin.time.measureTimedValue

abstract class AoCPuzzle {
    private val terminal = Terminal()

    private val name = this::class.simpleName!!
    private val day = name.substringAfter("Day").toInt()
    private val year = this::class.qualifiedName!!.split('.').dropLast(1).last().substringAfter("aoc").toInt()
    private val input: Input = run {
        val inputPath = Path("resources/aoc$year/$name.txt")
        if (inputPath.notExists()) downloadInput(day, year, terminal)
        inputPath.readLines()
    }

    protected open val part1Tests = listOf(Test())
    protected open val part2Tests = listOf(Test())

    protected open fun part1(input: Input): Number = 0
    protected open fun part2(input: Input): Number = 0

    protected fun testPart1() = part1Tests.testAll("Part 1 Test", ::part1)
    protected fun testPart2() = part2Tests.testAll("Part 2 Test", ::part2)

    private fun <T> List<Test>.testAll(name: String, action: (Input) -> T): Boolean =
        singleOrNull()?.render(name, action) ?: mapIndexed { index, test ->
            test.render(name + " ${index + 1}", action)
        }.all { it }

    protected fun runPart1() = render("Part 1", ::part1)
    protected fun runPart2() = render("Part 2", ::part2)

    class Test(val expected: Number = 0, val input: String = "") {
        val inputLines = input.lines()
    }

    private fun <T> render(name: String, action: (Input) -> T): T = terminal.run {
        println(HorizontalRule(cyan(name), ruleCharacter = "═", ruleStyle = TextStyle(cyan)))
        renderTimed(bold.style) { action(input) }
    }

    private fun <T> Test.render(name: String, action: (Input) -> T): Boolean =
        with(terminal) {
            println(HorizontalRule(yellow(name), ruleStyle = TextStyle(yellow)))
            val result = action(inputLines)
            val passed = result == expected
            if (passed) {
                println((brightGreen + bold)("Passed"))
            } else {
                println(grid {
                    style = brightRed
                    column(0) {
                        width = ColumnWidth.Expand(1)
                    }
                    column(1) {
                        width = ColumnWidth.Expand(1)
                        align = TextAlign.RIGHT
                    }
                    column(2) {
                        width = ColumnWidth.Expand(2)
                    }
                    row {
                        cell(bold("Failed")) {
                            rowSpan = 2
                        }
                        cell("actual:")
                        cell(bold(result.toString()))
                    }
                    row {
                        cell("expected:")
                        cell(bold(expected.toString()))
                    }
                })
            }
            passed
        }

    protected fun <R> renderTimed(style: TextStyle = reset.style, action: () -> R): R {
        val (value, duration) = measureTimedValue(action)
        terminal.println(horizontalLayout {
            whitespace = Whitespace.PRE_WRAP
            column(0) {
                width = ColumnWidth.Expand()
            }
            column(1) {
                width = ColumnWidth.Auto
                padding = Padding { left = 1 }
                align = TextAlign.RIGHT
            }
            cell(style(value.toString()))
            cell(gray(duration.toString()))
        })
        return value
    }

    protected fun println(message: Any?, style: TextStyle = gray) = terminal.println(
        style(message.toString()),
        whitespace = Whitespace.PRE_WRAP
    )
    protected fun <T> T.println(style: TextStyle = gray): T = also { println(this, style) }
}

File: rust/src/util/aocpuzzle.rs
use std::fs;
use std::path::Path;
use std::str::Lines;
use std::time::Instant;
use colored::Colorize;
use crate::util::input::download_input;

const WIDTH: usize = 80;

pub struct Test {
    pub expected: usize,
    pub input: String
}

impl Test {
    pub fn new(expected: usize, input: String) -> Self {
        Self {
            expected,
            input: input.trim().to_string(),
        }
    }
}

pub trait AOCPuzzle<const DAY: usize, const YEAR: usize> {
    fn part1_tests(&self) -> Vec<Test>;
    fn part1(&self, input: Lines<'_>) -> usize;

    fn part2_tests(&self) -> Vec<Test>;
    fn part2(&self, input: Lines<'_>) -> usize;
}

pub struct AOCPuzzleRunner<P: AOCPuzzle<DAY, YEAR>, const DAY: usize, const YEAR: usize> {
    pub puzzle: P,
    input: String
}

impl<P: AOCPuzzle<DAY, YEAR>, const DAY: usize, const YEAR: usize> AOCPuzzleRunner<P, DAY, YEAR> {
    pub fn new(puzzle: P) -> Self {
        let filename = format!("../resources/aoc{}/Day{:02}.txt", YEAR, DAY);
        let path = Path::new(&filename);
        if !path.exists() {
            download_input(YEAR.try_into().unwrap(), DAY as i32);
        }
        let input = fs::read_to_string(path).unwrap();
        Self {
            puzzle,
            input,
        }
    }

    pub fn test_part1(&self) -> bool {
        self.test_all(&self.puzzle.part1_tests(), "Part 1 Test", |this, input| this.puzzle.part1(input))
    }

    pub fn test_part2(&self) -> bool {
        self.test_all(&self.puzzle.part2_tests(), "Part 2 Test", |this, input| this.puzzle.part2(input))
    }

    fn test_all(&self, tests: &Vec<Test>, name: &str, action: fn(&Self, Lines<'_>) -> usize) -> bool {
        if tests.len() == 1 {
            self.render_test(name, tests.first().unwrap(), action)
        } else  {
            tests.iter().enumerate().fold(true, move |value, (index, test)| {
                let name = format!("{} {}", name, index + 1);
                self.render_test(&name, test, action) && value
            })
        }
    }

    pub fn run_part1(&self) -> usize {
        self.render("Part 1", |this, input| this.puzzle.part1(input))
    }

    pub fn run_part2(&self) -> usize {
        self.render("Part 2", |this, input| this.puzzle.part2(input))
    }

    fn render(&self, name: &str, action: fn(&Self, Lines<'_>) -> usize) -> usize {
        println!("{}", self.labeled_horizontal_rule('=', name).cyan());

        self.render_timed(|| action(self, self.input.lines()))
    }

    fn render_test(&self, name: &str, test: &Test, action: fn(&Self, Lines<'_>) -> usize) -> bool {
        println!("{}", self.labeled_horizontal_rule('-', name).yellow());

        let result = action(self, test.input.lines());
        let passed = result == test.expected;
        if passed {
            println!("{}", "Passed".bright_green().bold())
        } else {
            println!("{:<width1$}{:>width1$}{:<width2$}", "Failed".red().bold(), "actual: ".red(), format!("{}", result).red().bold(), width1 = WIDTH / 4, width2 = WIDTH / 2);
            println!("{:>width$}{:<width$}", "expected: ".red(), test.expected.to_string().red().bold(), width = WIDTH / 2);
        }
        passed
    }

    fn render_timed(&self, action: impl Fn() -> usize) -> usize {
        let start = Instant::now();
        let result = action();
        let duration = start.elapsed();

        println!("{:<width$}{:>width$?}", result.to_string().bold(), duration, width = WIDTH / 2);

        result
    }

    fn horizontal_rule(&self, rule_char: char) -> String {
        rule_char.to_string().repeat(WIDTH)
    }

    fn labeled_horizontal_rule(&self, rule_char: char, label: &str) -> String {
        let rule = self.horizontal_rule(rule_char);
        format!("{}{}{}", &rule[0..(WIDTH - label.len()) / 2], label, &rule[(WIDTH + label.len()) / 2..WIDTH])
    }
}